android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/routing/backstack/BackStackManager.kt

```diff
@@ -0,0 +1,187 @@
+        val backStack: List<BackStackElement<C>> = emptyList()
+    ) : Parcelable {
+        val current: C
+            get() = backStack.last().configuration
+
+        val canPop: Boolean
+            get() = backStack.size > 1
+    }
+
+    class BooststrapperImpl<C : Parcelable>(
+        private val state: State<C>,
+        private val initialConfiguration: C
+    ) : Bootstrapper<Action<C>> {
+        override fun invoke(): Observable<Action<C>> = when {
+            state.backStack.isEmpty() -> just(Execute(NewRoot(initialConfiguration)))
+            else -> just(ReattachRibsOfLastEntry())
+        }
+    }
+
+    sealed class Wish<C : Parcelable> {
+        data class Replace<C : Parcelable>(val configuration: C) : Wish<C>()
+        data class Push<C : Parcelable>(val configuration: C) : Wish<C>()
+        data class NewRoot<C : Parcelable>(val configuration: C) : Wish<C>()
+        class Pop<C : Parcelable> : Wish<C>()
+        class ShrinkToBundles<C : Parcelable> : Wish<C>()
+        class TearDown<C : Parcelable> : Wish<C>()
+    }
+
+    sealed class Action<C : Parcelable> {
+        data class Execute<C : Parcelable>(val wish: Wish<C>) : Action<C>()
+        class ReattachRibsOfLastEntry<C : Parcelable> : Action<C>()
+    }
+
+    sealed class Effect<C : Parcelable> {
+        data class Replace<C : Parcelable>(val newEntry: BackStackElement<C>) : Effect<C>()
+        data class Push<C : Parcelable>(val updatedOldEntry: BackStackElement<C>, val newEntry: BackStackElement<C>) : Effect<C>()
+        data class NewRoot<C : Parcelable>(val newEntry: BackStackElement<C>) : Effect<C>()
+        data class Pop<C : Parcelable>(val updatedOldEntry: BackStackElement<C>) : Effect<C>()
+        data class UpdateBackStack<C : Parcelable>(val updatedBackStack: List<BackStackElement<C>>) : Effect<C>()
+    }
+
+    class ActorImpl<C : Parcelable>(
+        private val connector: BackStackRibConnector<C>
+    ) : Actor<State<C>, Action<C>, Effect<C>> {
+
+        override fun invoke(state: State<C>, action: Action<C>): Observable<out Effect<C>> =
+            when (action) {
+                is ReattachRibsOfLastEntry -> Observable.defer {
+                    return@defer just(
+                        connector.enter(
+                            state.backStack.last()
+                        )
+                    ).flatMap { empty<Effect<C>>() }
+                }
+
+                is Execute -> {
+                    val wish = action.wish
+                    when (wish) {
+                        is Replace -> when {
+                            wish.configuration != state.backStack.last() ->
+                                connector.switchToNew(
+                                    state.backStack,
+                                    wish.configuration,
+                                    detachStrategy = DESTROY
+                                ).flatMap {
+                                    just(Effect.Replace(it.second))
```

Comment:
Here and later:
1) if you really want to use a `Pair`, please use destructuring to name parameters
e.g `.flatMap { (_, toReplace) -> just(Effect.Replace(toReplace)) }
2) can we use `.map { Effect }` instead of `.flatMap { just(Effect) }` ?